src/main/scala/utils/Solver.scala

```diff
@@ -223,23 +224,29 @@ object Solver {
           // finds the two connections needed to add the neighbor to the copied graph
           val connectionA = connections.getConnection(source, neighbor)
           val connectionB = connections.getConnection(neighbor, target)
-          copy.addConnection(connectionA)
-          copy.addConnection(connectionB)
 
-          // calculating the added weight
-          val addedWeight: Quantity[Length] = connectionA.distance
-            .add(connectionB.distance)
-            .subtract(removedEdge.getDistance)
+          (connectionA, connectionB) match {
+            case (Some(conA), Some(conB)) =>
+              copy.addConnection(conA)
+              copy.addConnection(conB)
 
-          Some(
-            StepResultOption(
-              copy,
-              neighbor,
-              List(connectionA, connectionB),
-              removedEdge,
-              addedWeight
-            )
-          )
+              // calculating the added weight
+              val addedWeight: Quantity[Length] = conA.distance
+                .add(conB.distance)
+                .subtract(removedEdge.getDistance)
+
+              Some(
+                StepResultOption(
+                  copy,
+                  neighbor,
+                  List(conA, conB),
+                  removedEdge,
+                  addedWeight
+                )
+              )
+            case _ =>
+              None
```

Comment:
Wouldn't this be a reason to throw an Exception? Since we want to insert the neighbour ( BE ;) )  but it didn't worked out somehow?

Reply:
A neighbour could be any `node` that is not yet connected to the graph. The only possible way a problem could occur would be if we have not connected `nodes` and can't connect at least one of them.

If we throw an exception here, we would have to return a `Try`. And later check if there is at least one `Success`.

What do you think would be the better solution.

Reply:
Do we expect the edge to always exist? If not, it's not an error and `None` should be fine, right?

Reply:
Good question. To be precise, a `Connection` don't mean that two `nodes` are **directly** connected. It only means that there is at least some way to connect them, because we assume that the next `node` we connect, is the one that adds the least extra length to the graph.

Reply:
Ok, let me try to check if I understand this right. `connections.getConnection` will return an empty option for a neighbour node `new` in case there is no connection possible to either `source` or `target`. Only if this happens from all edges, this node `new` can't be connected, right? So for every single attempt to connect it, a `None` is fine. But do we cover the case where this happens for all the edges?

Reply:
For now the algorithm stops if no new `node` can be connected. If we still have unconnected `nodes` a log message is logged.

Reply:
Fine for me, thanks for clarifying. If @sebastian-peter has no further questions this can be merged after resolving the conflicts.